### PR TITLE
Status bar: fix book progress

### PR DESCRIPTION
**Pull request by hius07**
Show the whole book progress percentage even when the chapter progress bar is enabled.
(Also deduplicated building of the settings menu items.)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11885)

### CODE REVIEW

File: frontend/apps/reader/modules/readerfooter.lua [hunk 1]
```diff
@@ -323,7 +324,7 @@ local footerTextGeneratorMap = {
         if prefix then
             string_percentage = prefix .. " " .. string_percentage
         end
-        return string_percentage:format(footer.progress_bar.percentage * 100)
+        return string_percentage:format(footer.pageno / footer.pages * 100)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:327]: Looks like you may be losing the case & logic for/of the first branch here:
https://github.com/koreader/koreader/blob/f0f37e3153ad669ddcbcd1fa067533c99938ddb0/frontend/apps/reader/modules/readerfooter.lua#L2177-L2185

Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:327]: Logic that I borrowed to use for the top status bar in https://github.com/koreader/koreader/commit/f00a88aef7e31c966e4e3e23e8dd3ccb54b0a2db

Comment (hius07) [reply frontend/apps/reader/modules/readerfooter.lua]: Thanks

File: frontend/apps/reader/modules/readerfooter.lua [hunk 2]
```diff
@@ -1947,7 +1696,7 @@ With this feature enabled, the current page is factored in, resulting in the cou
         table.insert(configure_items_sub_table , 5, {
             text_func = function()
                 if self.settings.battery_hide_threshold <= (Device:hasAuxBattery() and 200 or 100) then
-                    return T(_("Hide battery item when higher than: %1".. "%"), self.settings.battery_hide_threshold)
+                    return T(_("Hide battery item when higher than: %1".. " %"), self.settings.battery_hide_threshold)
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerfooter.lua:1699]: All these gaps are ugly.

`50%` looks clearly better.
`50 %` why???

File: frontend/apps/reader/modules/readerfooter.lua [hunk 3]
```diff
@@ -1347,14 +1302,13 @@ function ReaderFooter:addToMainMenu(menu_items)
             },
             {
                 text_func = function()
-                    return T(_("Minimum progress bar width: %1".. "%"), self.settings.progress_bar_min_width_pct)
+                    return T(_("Minimum progress bar width: %1".. " %"), self.settings.progress_bar_min_width_pct)
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerfooter.lua:1305]: Noooo

File: frontend/apps/reader/modules/readerfooter.lua [hunk 4]
```diff
@@ -1670,126 +1552,33 @@ With this feature enabled, the current page is factored in, resulting in the cou
...
+                    self:genItemSeparatorMenuItems("none"),
                 },
             },
             {
                 text = _("Item max width"),
                 sub_item_table = {
                     {
                         text_func = function()
-                            return T(_("Book-title item: %1".. "%"), self.settings.book_title_max_width_pct)
+                            return T(_("Book-title item: %1".. " %"), self.settings.book_title_max_width_pct)
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerfooter.lua:1579]: You get the point

File: frontend/apps/reader/modules/readerfooter.lua [hunk 5]
```diff
@@ -1813,10 +1602,9 @@ With this feature enabled, the current page is factored in, resulting in the cou
                     },
                     {
                         text_func = function()
-                            return T(_("Chapter-title item: %1".. "%"), self.settings.book_chapter_max_width_pct)
+                            return T(_("Chapter-title item: %1".. " %"), self.settings.book_chapter_max_width_pct)
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerfooter.lua:1605]: Another one

**Review by Commodore64user — COMMENTED** [review]
Messages are not in order, sorry. In essence, please don't add the spaces in between numbers and percentages. They are ugly 😬

**Comment by hius07** [comment]
> please don't add the spaces in between numbers and percentages

It was discussed and added to our design guide.

**Comment by Commodore64user** [comment]
Well KOReader is written in english, and should therefore use the rules of that language, i.e no space in between. Also, no body mentioned it when i switched it with the previous PR which in my book meant. We accept it.

**Comment by Frenzie** [comment]
|||
|---|---|
|no space| 50% |
|thin space| 50 % |
|hair space| 50 % |
|space| 50 % |

> Also, no body mentioned it when i switched it with the previous PR which in my book meant. We accept it.

If there was such a change we missed it because it was too much all at once. :-)

I don't recall this discussion. Would you be able to find it @hius07?

**Comment by hius07** [comment]
It was proposed by @zwim when adding units to the SpinWidget and accepted.

**Comment by hius07** [comment]
French modules
https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/frontend/apps/reader/modules/readerfont.lua#L771
https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/frontend/ui/data/css_tweaks.lua#L1034
Austrian modules
https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/plugins/coverimage.koplugin/main.lua#L640-L641
https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/plugins/autowarmth.koplugin/main.lua#L977

**Comment by Frenzie** [comment]
Thanks, so that was in https://github.com/koreader/koreader/pull/9046

I disagree that without a space is quite *the* standard in English, but it is more common. However… that's referring to within text while here we're talking primarily in isolation, which makes it math more than English.

I'm inclined to think it looks best with a hair space, but my only real preference is to be consistent where possible.

**Comment by Commodore64user** [comment]
> French modules
> 
> https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/frontend/apps/reader/modules/readerfont.lua#L771
> 
> https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/frontend/ui/data/css_tweaks.lua#L1034
> 
> Austrian modules
> 
> https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/plugins/coverimage.koplugin/main.lua#L640-L641
> 
> https://github.com/koreader/koreader/blob/81575ae24f4a53158c0f6f65c8e363f5fe5db8e3/plugins/autowarmth.koplugin/main.lua#L977

All of this is painful to see, my eyes...

**Comment by Commodore64user** [comment]
No space all the way.

**Comment by poire-z** [comment]
From a UX perspective - but not an English gramar one - some spacing makes it easier to pick the number `88` from the combo `88 %` - `88%` makes it harder, just because the `%` sign has quite the same body weight as digits, unlike `38°`.
(If using Unicode smaller spaces, I fear they could easily be eaten by translators - unless it feels it's the translator choice, but they probably won't notice some care/thinking/consistency is needed.)

**Comment by mergen3107** [comment]
I would vote for a space.
1) In scientific papers, for "percent" or other spelled-out units, it usually is with a space, like "50 Ohm", "100 Volt", "200 percent".
2) For units with short notation, it is "50 Ω", "100 V", so for percentage it should be consistent, like "200 %".

My 2 cents :D

**Review by poire-z — APPROVED** [review]

**Comment by mergen3107** [comment]
One more suggestion maybe that this white space should be non-breaking space (for other applications, where the units can overflow to second line, in status bar it is one line anyway)

**Comment by Commodore64user** [comment]
Apple also does it, the right way image

**Comment by hius07** [comment]
I think we can (step by step) replace SP% with NNBSP% everywhere.

**Comment by hius07** [comment]
But should we keep it translatable?
Maybe just append to the translated string?
https://github.com/koreader/koreader/blob/a082d85a0b8cf2ddaa39f4784d3b6b7a080be233/frontend/apps/reader/modules/readerfooter.lua#L1305

**Comment by Frenzie** [comment]
I don't know about bidi but it's also possible that Chinese prefers the full-width percent sign.

**Comment by NiLuJe** [comment]
> French

That might be the problem right there: French typography rules mandate ugly hairspaces before a whole lot of punctuation marks. (Too lazy to check if % is affected, because I hate that rule with a fiery passion anyway ;p).

**Pull request merged**